### ISSUE TITLE
Database-error creating entry using Moodle 3.11 and PostgreSQL 13

### ISSUE THREAD

**Issue by ksteitz** [coactum/moodle-mod_margic#4]
Plugin-installation and creating of a Margic-activity worked trouble-free and without errors. But after clicking the button "new entry" there is a database-error: "Error reading from database"

[25-Jan-2023 10:50:57 Europe/Berlin] Default exception handler: Fehler beim Lesen der Datenbank Debug: ERROR:  invalid input syntax for type bigint: ""
SELECT 'x' FROM mdl_margic_entries WHERE margic = $1 AND id = $2 LIMIT 1
[array (
  0 => '1',
  1 => '',
)]
Error code: dmlreadexception
* line 486 of /lib/dml/moodle_database.php: dml_read_exception thrown
* line 293 of /lib/dml/moodle_read_slave_trait.php: call to moodle_database->query_end()
* line 338 of /lib/dml/pgsql_native_moodle_database.php: call to pgsql_native_moodle_database->read_slave_query_end()
* line 892 of /lib/dml/pgsql_native_moodle_database.php: call to pgsql_native_moodle_database->query_end()
* line 2002 of /lib/dml/moodle_database.php: call to pgsql_native_moodle_database->get_recordset_sql()
* line 1987 of /lib/dml/moodle_database.php: call to moodle_database->record_exists_sql()
* line 1971 of /lib/dml/moodle_database.php: call to moodle_database->record_exists_select()
* line 96 of /mod/margic/edit.php: call to moodle_database->record_exists()

table "mdl_margic_entries" remains empty accordingly. 

![grafik](https://user-images.githubusercontent.com/1639438/215040462-83a757f5-a371-4e64-8b16-11ebe8e76f12.png)

Are there known issues using PostgreSQL? 
Is there anything that I can test?

systeminfo:
- Moodle 3.11
- PostgreSQL 13.9 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 8.5.0 20210514 (Red Hat 8.5.0-15), 64-bit
- PHP Version 7.4.33
- mod_margic: 1.2.7 (2023011600)
- plugin-settings default

thanks for support!
Klaus

**Comment by HarryBleckert** [comment]
Hey Klaus
My first guess was also that it could be a postgreSQL issue and maybe it is.
But the simple solution seems to be an error with the default value being a string for optional parameter entryid when the expected parameter isd of type PARAM_INT.
Simple fix seems to be changing line 43 of edit.php from:
  $entryid = optional_param('entryid', '', PARAM_INT);
 to
  $entryid = optional_param('entryid', 0, PARAM_INT);

See my issue report also from today
[Error in line 43 of edit.php: Wrong default value](https://github.com/coactum/moodle-mod_margic/issues/3)

**Comment by ksteitz** [comment]
Hey Harry,

thank you for the quick answer and solution, that does the trick: 
replacing '' by 0 in edit.php

Then creating of an entry works.

But the bad news is that further plugin-testing results directly in the next error. 
If i try to save an annotation (using annotations.php) i get again a db-error: "Error writing to database"

[27-Jan-2023 12:29:44 Europe/Berlin] Default exception handler: Fehler beim Schreiben der Datenbank Debug: ERROR:  syntax error at or near "end"
LINE 1: ...ontainer,endcontainer,startoffset,endoffset,start,end,exact,...
                                                             ^
INSERT INTO mdl_margic_annotations (margic,entry,userid,timecreated,timemodified,type,startcontainer,endcontainer,startoffset,endoffset,start,end,exact,prefix,suffix,text) VALUES($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16) RETURNING id
[array (
  'margic' => 4,
  'entry' => 4,
  'userid' => '18510',
  'timecreated' => 1674818984,
  'timemodified' => 0,
  'type' => '14',
  'startcontainer' => '/p[3]',
  'endcontainer' => '/p[3]',
  'startoffset' => 0,
  'endoffset' => 6,
  'start' => 47,
  'end' => 53,
  'exact' => 'Fehler',
  'prefix' => 'selbstHier könnte mehr stehen...',
  'suffix' => '',
  'text' => 'was soll das ausdrücken?',
)]
Error code: dmlwriteexception
* line 489 of /lib/dml/moodle_database.php: dml_write_exception thrown
* line 293 of /lib/dml/moodle_read_slave_trait.php: call to moodle_database->query_end()
* line 338 of /lib/dml/pgsql_native_moodle_database.php: call to pgsql_native_moodle_database->read_slave_query_end()
* line 1091 of /lib/dml/pgsql_native_moodle_database.php: call to pgsql_native_moodle_database->query_end()
* line 1139 of /lib/dml/pgsql_native_moodle_database.php: call to pgsql_native_moodle_database->insert_record_raw()
* line 198 of /mod/margic/annotations.php: call to pgsql_native_moodle_database->insert_record()

![grafik](https://user-images.githubusercontent.com/1639438/215078257-6f2ee873-8e01-4571-9fc8-ba45d9257611.png)


The plugin looks daunting buggy for me using pg at the moment; struggling from one db-error to the next...

**Comment by dtnolte** [comment]
Hi all,
thanks for your reports and sorry for the late reply.  I have investigated both errors and im working on a new plugin version to fix them. I also looked into the pull-request #6 by NinaHerrmann (many thanks for the wonderfull work!) but discovered two issues with it (e. g. that the way the db fields are renamed is not working for other database types). So i have to make some adjustments to it. When im ready with this and my tests against a moodle with a pq database are finished i will release a new version, hopefully in the next two weeks.
